internal/http/parser/http_response.go

```diff
@@ -0,0 +1,263 @@
+	Key   string
+	Value string
+}
+
+type StatusLine struct {
+	Version    string
+	StatusCode int
+	Reason     string
+}
+
+type HTTPResponse struct {
+	StatusLine StatusLine
+
+	Headers []Header
+
+	Body []byte
+}
+
+func Parse(data []byte) (httpResponse HTTPResponse, readBytesCount int, err error) {
+	reader := bytes.NewReader(data)
+
+	response, err := doParseResponse(reader)
+	if err != nil {
+		return HTTPResponse{}, 0, err
+	}
+
+	return response, len(data) - reader.Len(), nil
+}
+
+func parseStatusLine(reader *bytes.Reader) (StatusLine, error) {
+	var statusLine StatusLine
+
+	offsetBeforeCurrentSection := GetReaderOffset(reader)
+	versionB, err := ReadBytes(reader, 9)
+	if err == io.EOF {
+		return StatusLine{}, IncompleteHTTPResponseError{
+			Reader:  reader,
+			Message: fmt.Sprintf("Expected: HTTP-version, Received: %q", versionB),
+		}
+	}
+	version := string(versionB)
+	// HTTP/1.X
+	if version[:8] != "HTTP/1.1" {
+		reader.Seek(int64(offsetBeforeCurrentSection), io.SeekStart)
+		return StatusLine{}, InvalidHTTPResponseError{
+			Reader:  reader,
+			Message: fmt.Sprintf("Expected 'HTTP/1.1', Received: %q", version[:8]),
+		}
+	}
+	if version[len(version)-1] != ' ' {
+		reader.Seek(int64(offsetBeforeCurrentSection+9), io.SeekStart)
+		return StatusLine{}, InvalidHTTPResponseError{
+			Reader:  reader,
+			Message: "Expected white-space after 'HTTP/1.1'",
+		}
+	}
+	version = version[:8]
+	statusLine.Version = version
+
+	offsetBeforeCurrentSection = GetReaderOffset(reader)
+	statusB, err := ReadUntilAnyDelimiter(reader, [][]byte{SPACE, TAB})
+	if err == io.EOF {
+		reader.Seek(int64(-2), io.SeekCurrent)
+		return StatusLine{}, IncompleteHTTPResponseError{
+			Reader:  reader,
+			Message: "Expected reason phrase (example: 'OK' for 200 response) at end of status line",
+		}
+	}
+	statusCode := string(statusB)
+	if len(statusCode) != 3 {
+		reader.Seek(int64(offsetBeforeCurrentSection), io.SeekStart)
+		return StatusLine{}, InvalidHTTPResponseError{
+			Reader:  reader,
+			Message: "Invalid status-code field length, Expected: 3 digits, Received: " + strconv.Itoa(len(statusCode)),
+		}
+	}
+	intStatusCode, err := strconv.Atoi(statusCode)
+	if err != nil {
+		reader.Seek(int64(offsetBeforeCurrentSection), io.SeekStart)
+		return StatusLine{}, InvalidHTTPResponseError{
+			Reader:  reader,
+			Message: "Invalid status-code field, Expected integer value, Received: " + statusCode,
+		}
+	}
+	statusLine.StatusCode = intStatusCode
+
+	// Intentionally lax. rfc9112.html#section-4-8
+	reasonB, err := ReadUntilCRLF(reader)
+	if err == io.EOF {
+		return StatusLine{}, IncompleteHTTPResponseError{
+			Reader:  reader,
+			Message: "Expected CRLF after status line",
+		}
+	}
+	reason := string(reasonB)
+	statusLine.Reason = reason
+	return statusLine, nil
+}
+
+func parseHeaders(reader *bytes.Reader) ([]Header, error) {
+	allHeadersFound := false
+	headers := []Header{}
+
+	for !allHeadersFound {
+		offsetBeforeCRLF := GetReaderOffset(reader)
```

Comment:
```suggestion
		startOffset := GetReaderOffset(reader)
```

(doesn't this offset refer to the start of the header? The CRLF is at the end)